django/core/cache/backends/db.py

```diff
@@ -46,41 +45,44 @@ class CacheEntry(object):
 class DatabaseCache(BaseDatabaseCache):
 
     # This class uses cursors provided by the database connection. This means
-    # it reads expiration values as aware or naive datetimes depending on the
-    # value of USE_TZ. They must be compared to aware or naive representations
-    # of "now" respectively.
+    # it reads expiration values as aware or naive datetimes, depending on the
+    # value of USE_TZ and whether the database supports time zones. They must
+    # be compared to aware or naive representations of "now" respectively.
 
     # But it bypasses the ORM for write operations. As a consequence, aware
     # datetimes aren't made naive for databases that don't support time zones.
-    # We work around this problem by always using naive datetimes when writing
-    # expiration values, in UTC when USE_TZ = True and in local time otherwise.
+
+    # We work around this by calling directly the adapters or converters.
 
     def get(self, key, default=None, version=None):
         key = self.make_key(key, version=version)
         self.validate_key(key)
         db = router.db_for_read(self.cache_model_class)
-        table = connections[db].ops.quote_name(self._table)
+        connection = connections[db]
+        table = connection.ops.quote_name(self._table)
 
-        with connections[db].cursor() as cursor:
+        with connection.cursor() as cursor:
             cursor.execute("SELECT cache_key, value, expires FROM %s "
                            "WHERE cache_key = %%s" % table, [key])
             row = cursor.fetchone()
         if row is None:
             return default
-        now = timezone.now()
         expires = row[2]
-        if connections[db].features.needs_datetime_string_cast and not isinstance(expires, datetime):
-            # Note: typecasting is needed by some 3rd party database backends.
-            # All core backends work without typecasting, so be careful about
-            # changes here - test suite will NOT pick regressions here.
-            expires = typecast_timestamp(str(expires))
-        if expires < now:
+
+        if hasattr(connection.ops, 'convert_datetimefield_value'):
```

Comment:
I think this line is questionable, too. If I'm reading things correctly, in the current implementation the method name `convert_datetimefield_value` is not intended to have any magical inherent meaning, it's simply a method referenced by `get_db_converters`. This line changes that.

I _think_ that the fully correct approach here would be to construct an `Expression(output_field=DateTimeField())` and then call both`connection.ops.get_db_converters(expression)` and `expression.get_db_converters(connection)` to get the converters to run. Then you also have an expression to pass in for the second argument. The fourth argument (context) I think should be an empty dictionary rather than `None`.

This does point out the difficulty of writing cross-database-portable raw SQL under the new regime...